src/output.rs

```diff
@@ -17,10 +18,14 @@ pub enum Argument {
 }
 
 impl FilterPrototype {
-    pub fn new(name: &str, arguments: Vec<Argument>) -> FilterPrototype {
+    pub fn new(name: &str,
+               arguments: Vec<Argument>,
+               named_arguments: Vec<(String, Argument)>)
+               -> FilterPrototype {
         FilterPrototype {
             name: name.to_owned(),
             arguments: arguments,
+            named_arguments: named_arguments,
```

Comment:
> How can we make this easy for filter writers?

HashMap vs list of tuples vs some kind of flat map (list of tuples with a Map API) don't really matter except for performance as long as we provide a stable API on hitting 1.0.

For how to make it easier on filter writers, one idea I just had is for a filter to do the following
```rust
let required = ["param0", "param1"];
let optional = ["param2", "param3"];
check_args_len(args, required, optional)?;

let index = Index(args, required, optional);
let offset = index[0]...;
let length = index.get(1)...;
```

The Index object implements all of the lookup rules for the developer.  Positional arguments are passed in for fast access for positional arguments but if the positional argument isn't there, the `required` and `optional` lists are indexed to find the name to look up the value in `named_arguments`.

This actually gets me thinking. 
- Clients need to rely on our helper code to get the correct semantics.  This means we need to expose it and hope they use it
- By exposing helper code, we are increasing our compatibility surface

I wonder if what we should have filters register `required` and `optional` arrays along with their `name` and `fn`.  We then have several options
- Pass the `Index` object (probably renamed to `Arguments`) to a filter
- Merge `named_arguments` into the unnamed arguments, passing one `Vec` in.  We then need to deal with holes in optional arguments caused by named arguments (`opt0: val0, opt3: val3`, no `opt1` or `opt2` defined).  We could handle this by putting all arguments into an `Option` or passing two `Vec`s to a filter, one for required and one for optional parameters.